SYCL/ESIMD/api/functional/functions/functions_atomic_update_inc_dec.cpp

```diff
@@ -0,0 +1,250 @@
+    static_assert(NumElemsToChange <= NumElems);
+    bool passed = true;
+    constexpr DataT base_value = 100;
+    constexpr DataT result_fill_value = 5;
+
+    auto init_values =
+        functions::get_init_values<NumElemsToChange, ChosenOperator, DataT>(
+            base_value);
+    shared_allocator<DataT> allocator(queue);
+    shared_vector<DataT> shared_init_values(init_values.begin(),
+                                            init_values.end(), allocator);
+    shared_vector<DataT> shared_result(NumElems, result_fill_value, allocator);
+
+    auto offsets =
+        functions::get_offsets<NumElemsToChange, AlgorithmToChange>();
+    shared_vector<size_t> shared_offsets(offsets.begin(), offsets.end(),
+                                         shared_allocator<size_t>(queue));
+
+    constexpr int NumberIteractions = 16;
+    sycl::nd_range<NDRangeDim> nd_range =
+        get_sycl_nd_range<NDRangeDim>(NumberIteractions);
+
+    queue.submit([&](sycl::handler &cgh) {
+      DataT *shared_init_values_ptr = shared_init_values.data();
+      DataT *shared_result_ptr = shared_result.data();
+      size_t *const offsets_ptr = shared_offsets.data();
+
+      cgh.parallel_for<
+          Kernel<DataT, NumElems, TestCaseT, ChosenOperatorT, NDRangeDimT,
+                 NumElemsToChangeT, ChangeElemFilterT, AlgorithmToChangeT>>(
+          nd_range, [=](sycl::nd_item<1> nd_item) SYCL_ESIMD_KERNEL {
+            const size_t work_item_index = nd_item.get_global_linear_id();
+            TestCaseT::template call_esimd_function<
+                DataT, NumElems, NumElemsToChange, ChangeElemFilterT,
+                ChosenOperator>(shared_init_values_ptr, shared_result_ptr,
+                                offsets_ptr, work_item_index);
+          });
+    });
+    queue.wait_and_throw();
+
+    std::sort(shared_result.begin(), shared_result.end());
+
+    ChangeElemFilterT filter{};
+    size_t num_changed_elems = 0;
+    // Collect the number of elements that should be changed. This value will be
+    // used to calculate expected value for all elements that will be returned
+    // after atomic_update call the sum of changed elements.
+    for (size_t i = 0; i < NumElemsToChange; ++i) {
+      if (filter(i) == 1) {
+        ++num_changed_elems;
+      }
+    }
+
+    for (size_t i = 0; i < NumElems; ++i) {
+      const DataT &retrieved = shared_result[i];
+      if (i < NumberIteractions) {
+        if (result_fill_value != retrieved) {
+          passed = fail_test(i, result_fill_value, retrieved, data_type,
+                             "atomic_update return value", operator_type);
+        }
+      } else {
+        const DataT &expected = base_value * num_changed_elems +
+                                (i - NumberIteractions) * num_changed_elems;
+        if (expected != retrieved) {
+          passed = fail_test(i, expected, retrieved, data_type,
+                             "atomic_update return value", operator_type);
```

Comment:
Do we really:

- use `atomic_update` over something with `i >= NumberIterations`
- change `shared_result[i]` on device side in case of `i >= NumberIterations`?

Doesn't `shared_result` need only to be `NumberIterations` in size? Or maybe I'm missing something.

Reply:
This commit has been applied in this commit: 556d3dad42da4b4d9382fb3612f3051931ad6a84
`shared_result` now used `NumberIteractions` as vector size [here](https://github.com/intel/llvm-test-suite/pull/939/commits/556d3dad42da4b4d9382fb3612f3051931ad6a84#diff-afaea17121d6706af52229dcfb462795dfe86fde5c98f09b836505a741830b6bR112-R113)